resolwe/flow/managers/base.py

```diff
@@ -74,26 +168,365 @@ def discover_engines(self):
         return os.linesep.join(export_commands) + os.linesep + program
 
-    def run(self, data_id, script, priority='normal', run_sync=False, verbosity=1):
+    def run(self, data_id, dest_dir, argv, priority='normal', run_sync=False, verbosity=1):
         """Run process."""
-        raise NotImplementedError('`run` function not implemented')
+        raise NotImplementedError("`run` function not implemented")
+
+    def _get_per_data_dir(self, dir_base, data_id):
+        """Extend the given base directory with a per-data component.
+
+        The method creates a private path for the data object, such as::
+
+            ./test_data/1/
+
+        if ``base_dir`` is ``'./test_data'`` and ``data_id`` is ``1``.
+
+        :param dir_base: The base path to be extended.
+        :param data_id: The data object id used for the extending.
+        :return: The new path path for the data object.
+        :rtype: str
+        """
+        # Use django settings here, because the state must be preserved
+        # across events. This also implies the directory settings can't
+        # be patched outside the manager and then just sent along in the
+        # command packets.
+        result = self.settings_actual.get('FLOW_EXECUTOR', {}).get(dir_base, '')
+        return os.path.join(result, str(data_id))
+
+    def _prepare_data_dir(self, data_id):
+        """Prepare destination directory where the data will live.
+
+        :param data_id: The data object id for which to prepare the
+            private execution directory.
+        :return: The prepared data directory path.
+        :rtype: str
+        """
+        output_path = self._get_per_data_dir('DATA_DIR', data_id)
+        dir_mode = self.settings_actual.get('FLOW_EXECUTOR', {}).get('DATA_DIR_MODE', 0o755)
+        os.mkdir(output_path, mode=dir_mode)
+        # os.mkdir is not guaranteed to set the given mode
+        os.chmod(output_path, dir_mode)
+        return output_path
+
+    def _prepare_context(self, data_id, data_dir, runtime_dir, **kwargs):
+        """Prepare settings and constants jsons for the executor.
+
+        Settings and constants provided by other ``resolwe`` modules and
+        :class:`~django.conf.settings` are all inaccessible in the
+        executor once it is deployed, so they need to be serialized into
+        the runtime directory.
+
+        :param data_id: The data object id being prepared for.
+        :param data_dir: The target execution directory for this data
+            object.
+        :param runtime_dir: The target runtime support directory for
+            this data object; this is where the environment is
+            serialized into.
+        :param kwargs: Extra settings to include in the main settings
+            file.
+        """
+        files = {}
+
+        settings_dict = dict(kwargs)
+        settings_dict.update(self.settings_actual)
+        settings_dict['DATA_DIR'] = data_dir
```

Comment:
They're not overrides, precisely. `DATA_DIR` as it appears here doesn't appear anywhere else, including settings.py; it is what the manager decides will be the directory the executor must change to to run the process (this will usually be more or less directly what it says in `FLOW_EXECUTOR['DATA_DIR']`, but that's irrelevant). `settings_actual` aren't overrides either, it's a dictionary of the current state of django global settings at the point where the `communicate` call was made, including immediate overrides given to it.